Bump minSdk to 21 - Android 5 / Lollipop

litetex opened the pull request:
#### What is it?
- [ ] Bugfix (user facing)
- [ ] Feature (user facing)
- [x] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
* Bump minSdk to ``21`` - Android 5 / Lollipop - Dropped support for Android 4.4 and removed all corresponding code snippets

#### Fixes the following issue(s)
- Fixes #7602 

#### APK testing 
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR.

#### Due diligence
- [x] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

Fs00 commented:
There also resources in `values-v21` and `values-night-v21` folders that can be put into their corresponding files in `values` folder.

litetex commented:
@Fs00 

Thank you for the hint.
I consolidated the styles 😄

Fs00 commented:
Great! After a quick look at the code I found other things that are unnecessary on API 21+:
- [AndroidX Multidex library](https://github.com/TeamNewPipe/NewPipe/blob/e127697fff54163196cd76e393d59f7d2726601f/app/build.gradle#L211) and its relative `multiDexEnabled` config parameter are not needed anymore, since Android 5+ natively supports multiple DEX files
- [Vector drawables backwards-compatibility](https://developer.android.com/guide/topics/graphics/vector-drawable-resources#vector-drawables-backward-solution) (enabled [here](https://github.com/TeamNewPipe/NewPipe/blob/e127697fff54163196cd76e393d59f7d2726601f/app/build.gradle#L25)) can be disabled (vector are natively supported on Android 5+) and usages of attribute `app:srcCompat` need to be replaced with `android:src`

TobiGr commented:
Please also check whether we can disable `coreLibraryDesugraing`. That is crushing reporoducible / deterministic builds (see #6486).

Are you going to upgrade okHttp in this PR?

litetex commented:
Thank you guys for the hints, I will have a look.

> Are you going to upgrade okHttp in this PR?

No that should be a new one. I'm just dropping 4.4 here 😄

litetex commented:
> Please also check whether we can disable `coreLibraryDesugraing`. That is crushing reproducible / deterministic builds (see #6486).

This is not possible.
Otherwise the app crashes on devices < Android 8 / API 26 because Java 8 features are missing.

<details><summary>Fun Fact: Compiling works execution not</summary>

``` 
01/04 18:26:38: Launching 'app' on Pixel 3a API 22.
Install successfully finished in 28 s 40 ms.
$ adb shell am start -n "org.schabi.newpipe.debug.increaseminsdk/org.schabi.newpipe.MainActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Connected to process 5554 on device 'Pixel_3a_API_22 [emulator-5556]'.
Connected to process 5624 on device 'Pixel_3a_API_22 [emulator-5556]'.
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
I/ACRA: ACRA is enabled for org.schabi.newpipe.debug.increaseminsdk, initializing...
D/InfoCache: clearCache() called
I/art: Rejecting re-init on previously-failed class java.lang.Class<org.ocpsoft.prettytime.PrettyTime$$ExternalSyntheticLambda0>
    Rejecting re-init on previously-failed class java.lang.Class<org.ocpsoft.prettytime.PrettyTime$$ExternalSyntheticLambda0>
D/LeakCanary: Updated AppWatcher.config: Config(watchDurationMillis=10000)
I/stetho: Listening on @stetho_org.schabi.newpipe.debug.increaseminsdk_devtools_remote
D/LeakCanary: Updated LeakCanary.config: Config(dumpHeap=false)
D/AndroidRuntime: Shutting down VM
E/ACRA: ACRA caught a NoClassDefFoundError for org.schabi.newpipe.debug.increaseminsdk
    java.lang.NoClassDefFoundError: Failed resolution of: Ljava/time/format/DateTimeFormatter;
        at org.schabi.newpipe.error.ErrorActivity.<clinit>(ErrorActivity.java:72)
        at java.lang.reflect.Constructor.newInstance(Native Method)
        at java.lang.Class.newInstance(Class.java:1606)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1066)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2226)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387)
        at android.app.ActivityThread.access$800(ActivityThread.java:151)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5254)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "java.time.format.DateTimeFormatter" on path: DexPathList[[zip file "/data/app/org.schabi.newpipe.debug.increaseminsdk-1/base.apk"],nativeLibraryDirectories=[/vendor/lib64, /system/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at org.schabi.newpipe.error.ErrorActivity.<clinit>(ErrorActivity.java:72) 
        at java.lang.reflect.Constructor.newInstance(Native Method) 
        at java.lang.Class.newInstance(Class.java:1606) 
        at android.app.Instrumentation.newActivity(Instrumentation.java:1066) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2226) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387) 
        at android.app.ActivityThread.access$800(ActivityThread.java:151) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5254) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698) 
    	Suppressed: java.lang.ClassNotFoundException: java.time.format.DateTimeFormatter
        at java.lang.Class.classForName(Native Method)
        at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
        at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
        		... 16 more
     Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available
Connected to process 5760 on device 'Pixel_3a_API_22 [emulator-5556]'.
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
I/ACRA: ACRA is enabled for org.schabi.newpipe.debug.increaseminsdk, initializing...
D/InfoCache: clearCache() called
I/art: Rejecting re-init on previously-failed class java.lang.Class<org.ocpsoft.prettytime.PrettyTime$$ExternalSyntheticLambda0>
    Rejecting re-init on previously-failed class java.lang.Class<org.ocpsoft.prettytime.PrettyTime$$ExternalSyntheticLambda0>
D/LeakCanary: Updated AppWatcher.config: Config(watchDurationMillis=10000)
I/stetho: Listening on @stetho_org.schabi.newpipe.debug.increaseminsdk_devtools_remote
D/LeakCanary: Updated LeakCanary.config: Config(dumpHeap=false)
D/AndroidRuntime: Shutting down VM
E/ACRA: ACRA caught a NoClassDefFoundError for org.schabi.newpipe.debug.increaseminsdk
    java.lang.NoClassDefFoundError: Failed resolution of: Ljava/time/format/DateTimeFormatter;
        at org.schabi.newpipe.error.ErrorActivity.<clinit>(ErrorActivity.java:72)
        at java.lang.reflect.Constructor.newInstance(Native Method)
        at java.lang.Class.newInstance(Class.java:1606)
        at android.app.Instrumentation.newActivity(Instrumentation.java:1066)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2226)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387)
        at android.app.ActivityThread.access$800(ActivityThread.java:151)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5254)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "java.time.format.DateTimeFormatter" on path: DexPathList[[zip file "/data/app/org.schabi.newpipe.debug.increaseminsdk-1/base.apk"],nativeLibraryDirectories=[/vendor/lib64, /system/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at org.schabi.newpipe.error.ErrorActivity.<clinit>(ErrorActivity.java:72) 
        at java.lang.reflect.Constructor.newInstance(Native Method) 
        at java.lang.Class.newInstance(Class.java:1606) 
        at android.app.Instrumentation.newActivity(Instrumentation.java:1066) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2226) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387) 
        at android.app.ActivityThread.access$800(ActivityThread.java:151) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5254) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698) 
    	Suppressed: java.lang.ClassNotFoundException: java.time.format.DateTimeFormatter
        at java.lang.Class.classForName(Native Method)
        at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
        at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
        		... 16 more
     Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available
```

</details>

Fs00 commented:
@litetex It just came to my mind that if you disable vector drawables backwards compatibility, you need also to replace `app:drawableXXXCompat` attributes with their `android:drawableXXX` counterpart (I've seen that are some usages of those in NewPipe).

litetex commented:
> @litetex It just came to my mind that if you disable vector drawables backwards compatibility, you need also to replace `app:drawableXXXCompat` attributes with their `android:drawableXXX` counterpart (I've seen that there are some usages of those in NewPipe).

When I don't replace the `app:drawableXXXCompat` the app works as expected and there is also no warning inside Android studio.

~Could you add a reference where it is stated that you should rather use `android:drawableXXX` instead of `app:drawableXXXCompat`?~
→ https://stackoverflow.com/questions/63479754/getting-appdrawableendcompat-instead-of-androiddrawableend-warning-for-lollipo

Anyway, I replaced the usages now.

Plugdemholes-Imeanpatchdembugs commented:
Will there be a new seperate downloadable Legacy release for the last version that supported Android 4.4?
Would be a shame if this new Legacy release would just replace the old one, considering the last Legacy release for Android 4.1 still seems to work.

Asking because when I'm checking the changelog and the version of the current Legacy release, there doesn't seem to be any indication in the changelog that there was a change in minSdk in version 0.20.9 ?.?

opusforlife2 commented:
This will only be required once a security issue is found in OkHttp 3.x, right? What I mean is, could we keep releasing new versions with Kitkat support until we are forced to upgrade OkHttp to 4.x?

Stypox commented:
@opusforlife2 since OkHttp 3.x is now not supported anymore, it's just a matter of time before a vulnerability is found, and it would be difficult to *know* that a new vulnerability was found. So it is a good idea imo to just drop support for 4.4, upgrade to OkHttp 4.x and stay on the safe side

TacoTheDank commented:
This is indeed a cause for celebration!

I can think of a few other compatibility things that can be removed with this, but those aren't a priority. I can just create a PR for those if we do decide to go forward with dropping all legacy support (everything < 21).

opusforlife2 commented:
> So it is a good idea imo to just drop support for 4.4, upgrade to OkHttp 4.x and stay on the safe side

In that case, considering that 19 PRs are already merged for the next version, maybe we should only merge this PR and release ASAP.

SameenAhnaf commented:
Off-topic but I think, https://github.com/TeamNewPipe/NewPipe/issues/7009 will be reconsidered before merging this issue. At least, old users can hope to continue using Newpipe.

TacoTheDank commented on app/src/main/java/org/schabi/newpipe/error/ReCaptchaActivity.java:
```diff
@@ -108,11 +107,8 @@ public void onPageFinished(final WebView view, final String url) {
-            cookieManager.removeAllCookie();
-        }
+        cookieManager.removeAllCookies(value -> {
+        });
```
I don't believe this callback actually does anything. I think it can just be replaced with `cookieManager.removeAllCookies(null)`.

litetex replied:
done

Stypox commented on app/src/main/java/org/schabi/newpipe/error/ReCaptchaActivity.java:
```diff
@@ -107,12 +106,7 @@ public void onPageFinished(final WebView view, final String url) {
-        } else {
-            cookieManager.removeAllCookie();
-        }
+        CookieManager.getInstance().removeSessionCookies(null);
```
Why you now remove only session cookies ([i.e.](https://developer.android.com/reference/android/webkit/CookieManager#removeSessionCookies(android.webkit.ValueCallback%3Cjava.lang.Boolean%3E)) *cookies without an expiration date*), instead of all cookies with [`removeAllCookies(null)`](https://developer.android.com/reference/android/webkit/CookieManager#removeAllCookies(android.webkit.ValueCallback%3Cjava.lang.Boolean%3E))?

litetex replied:
Looks like I selected the wrong method in the autocomplete 😅

Stypox commented on app/src/main/res/layout/fragment_instance_list.xml:
```diff
@@ -46,6 +46,6 @@
         android:focusable="true"
         app:backgroundTint="?attr/colorPrimary"
         app:fabSize="auto"
-        app:srcCompat="@drawable/ic_add" />
```
I think this file can be reformatted with `Ctrl+Shift+Alt+L`, since the order of attributes here is not alphabetical (maybe other files, too)

Stypox commented on app/src/main/res/layout/list_comments_item.xml:
```diff
@@ -94,7 +94,7 @@
+        android:src="@drawable/ic_heart" />
 
 
     <!--we can uncomment below code if we need dislike button and count in future-->
```
I would also remove this. If in the future we need the dislike button, we will just copy-paste the XML for the like button, there is no need to keep some commented code that is probably old.

Stypox commented on app/src/main/res/layout/list_comments_mini_item.xml:
```diff
@@ -64,7 +64,7 @@
         android:layout_marginLeft="12dp"
         android:layout_toRightOf="@id/detail_thumbs_up_count_view"
         android:contentDescription="@string/detail_dislikes_img_view_description"
-        app:srcCompat="?attr/thumbs_down"
```
Same here

Stypox commented on app/src/main/res/values/styles_services.xml:
```diff
@@ -1,32 +1,38 @@
     <style name="LightTheme.SoundCloud" parent="LightTheme" tools:keep="@style/LightTheme_SoundCloud">
         <item name="colorPrimary">@color/light_soundcloud_primary_color</item>
-        <item name="colorPrimaryDark">@color/light_soundcloud_dark_color</item>
+        <item name="colorPrimaryDark">@color/light_soundcloud_primary_color</item>
```
(reminder for me to make sure the status bar is still darker in dark theme)

Stypox replied:
Yes, there is a problem:
<img src="https://user-images.githubusercontent.com/36421898/156531908-f7b3b6d8-ebe9-4db8-8234-02e565824cea.jpg" height="300px"/>

TacoTheDank replied:
I don't see why that needed to be changed in the first place? 🤔

litetex replied:
Is this a problem I should work on or just a sidenote.
Because if it's a problem I should work on, I don't know what is meant with the above picture. Everything works fine when I install the app:
<img src="https://user-images.githubusercontent.com/40789489/156818982-3b42b0d6-9285-4073-ae93-53cfa09bcbde.png" height=400 />
This is exactly the same behavior as with the current release.

Does this only occur on a specific android version?

litetex replied:
@Stypox Should be fixed now 😄

Stypox replied:
Can confirm

Stypox reviewed (changes requested):
Looks mostly good to me

litetex commented:
@Stypox Should be fixed now 😄

Stypox reviewed (approved):
Looks good to me. I smoke tested on my API 29 phone and on API 22 emulated and found no issue.

opusforlife2 commented:
I'm assuming we'll do this in a separate version bump? The current dev already has a bunch of PRs merged. Or does it not matter?

TacoTheDank commented:
@opusforlife2 Perhaps we can wait? Idk.

TacoTheDank commented:
@litetex Could I rebase this PR to fix the conflicts?

litetex commented:
@TacoTheDank Feel free to do so but note that this PR was create some months ago.
There are also certainly other things that need to be adjusted.

litetex reviewed (commented):
@TacoTheDank 
Are you finished with your changes?

If there is nothing additional to do I think we can merge the branch.

opusforlife2 commented:
Are we done with all the stuff that was proposed to be merged before removing Kitkat support? I can't exactly remember what those things were, though I think DASH and new stream notifications were two of them.

Stypox commented:
@opusforlife2 yes, we are done; there has been a recent discussion about whether to also wait for tex's DASH-2, but we decided not to wait any longer.

Stypox reviewed (approved):
Took a look at the code, lgtm, thank you to everyone involved!

leetfin commented:
Very sad to see this happen.

TacoTheDank commented:
> @TacoTheDank Are you finished with your changes?
> 
> If there is nothing additional to do I think we can merge the branch.

Yep, all done!

Stypox commented:
I tested some misc things and could not find issues

opusforlife2 commented:
> about whether to also wait for tex's DASH-2, but we decided not to wait any longer.

Isn't that purely the extractor? Extractor changes shouldn't have any influence on Legacy, no? I was under the impression that the front-end is what differs between NP and Legacy, which is why we wanted to add some critical front-end PRs before bumping.

Is that not the case?
